clover.rb

```diff
@@ -3,38 +3,688 @@
+      return @is_api if defined?(@is_api)
+      @is_api = env["HTTP_HOST"]&.start_with?("api.") || env["PATH_INFO"].start_with?("/api")
+    end
+  end
+  # :nocov:
+
+  def runtime?
+    return @is_runtime if defined?(@is_runtime)
+    @is_runtime = env["PATH_INFO"].start_with?("/runtime")
+  end
+
+  def web?
+    return @is_web if defined?(@is_web)
+    @is_web = !api? && !runtime?
+  end
+
+  def get_jwt_payload(request)
+    return unless (v = request.env["HTTP_AUTHORIZATION"])
+    jwt_token = v.sub(%r{\ABearer:?\s+}, "")
+    begin
+      JWT.decode(jwt_token, Config.clover_runtime_token_secret, true, {algorithm: "HS256"})[0]
+    rescue JWT::DecodeError
+    end
+  end
+
+  def csrf_tag(*)
+    render("components/form/hidden", locals: {name: csrf_field, value: csrf_token(*)})
+  end
+
+  def redirect_back_with_inputs
+    flash["old"] = request.params
+    request.redirect env["HTTP_REFERER"] || "/"
+  end
+
+  def has_project_permission(actions)
+    @project_permissions.intersection(Authorization.expand_actions(actions)).any?
+  end
+
+  def current_account
+    return @current_account if defined?(@current_account)
+    @current_account = Account[rodauth.session_value]
+  end
+
+  def json_params
+    @params ||= api? ? request.body.read : request.params.reject { _1 == "_csrf" }.to_json
+  end
+
+  # Assign some HTTP response codes to common exceptions.
+  def parse_error(e)
+    case e
+    when Sequel::ValidationFailed
+      code = 400
+      type = "InvalidRequest"
+      message = e.to_s
+    when CloverError
+      code = e.code
+      type = e.type
+      message = e.message
+      details = e.details
+    else
+      Clog.emit("route exception") { Util.exception_to_hash(e) }
+
+      code = 500
+      type = "UnexceptedError"
+      message = "Sorry, we couldn’t process your request because of an unexpected error."
+    end
+
+    response.status = code
+
+    {
+      code: code,
+      type: type,
+      message: message,
+      details: details
+    }
+  end
+
+  def fetch_location_based_prices(*resource_types)
+    # We use 1 month = 672 hours for conversion. Number of hours
+    # in a month changes between 672 and 744, We are  also capping
+    # billable hours to 672 while generating invoices. This ensures
+    # that users won't see higher price in their invoice compared
+    # to price calculator and also we charge same amount no matter
+    # the number of days in a given month.
+    BillingRate.rates.filter { resource_types.include?(_1["resource_type"]) }
+      .group_by { [_1["resource_type"], _1["resource_family"], _1["location"]] }
+      .map { |_, brs| brs.max_by { _1["active_from"] } }
+      .each_with_object(Hash.new { |h, k| h[k] = h.class.new(&h.default_proc) }) do |br, hash|
+      hash[br["location"]][br["resource_type"]][br["resource_family"]] = {
+        hourly: br["unit_price"].to_f * 60,
+        monthly: br["unit_price"].to_f * 60 * 672
+      }
+    end
+  end
+  plugin :not_found do
```

Comment:
General question for Roda: Can the not found exception be caught in the 'error_handler' when we aren't using the 'not_found' plugin? It might be more clear to handle all error cases in one place

Reply:
The `error_handler` is only called for exceptions.  `not_found` handles cases where the response uses status code 404 and has not written anything to the body.  So you cannot use `error_handler` for `not_found` responses.